packages/providers/src.ts/base-provider.ts

```diff
@@ -656,24 +681,194 @@ export class BaseProvider extends Provider {
+            this._events = this._events.filter((event) => {
+                if (event.tag !== eventTag) { return true; }
+                stopped.push(event);
+                return false;
+            });
+        }
+
+        stopped.forEach((event) => { this._stopEvent(event); });
         return this;
     }
+
+    get polling(): boolean {
+        return (this._poller != null);
+    }
+
+    set polling(value: boolean) {
+        if (value && !this._poller) {
+            this._poller = setInterval(() => { this.poll(); }, this.pollingInterval);
+
+            if (!this._bootstrapPoll) {
+                this._bootstrapPoll = setTimeout(() => {
+                    this.poll();
+
+                    // We block additional polls until the polling interval
+                    // is done, to prevent overwhelming the poll function
+                    this._bootstrapPoll = setTimeout(() => {
+                        // If polling was disabled, something may require a poke
+                        // since starting the bootstrap poll and it was disabled
+                        if (!this._poller) { this.poll(); }
+
+                        // Clear out the bootstrap so we can do another
+                        this._bootstrapPoll = null;
+                    }, this.pollingInterval);
+                }, 0);
+            }
+
+        } else if (!value && this._poller) {
+            clearInterval(this._poller);
+            this._poller = null;
+        }
+    }
+
+    async poll(): Promise<void> {
+        const pollId = nextPollId++;
+        // cleanup on more than 100 events
+        if (Object.keys(this._emittedEvents).length > 100) {
+            this.purgeOldEvents();
+        }
+        // Track all running promises, so we can trigger a post-poll once they are complete
+        const runners: Array<Promise<void>> = [];
+
+        // Emit a poll event with the current timestamp
+        this.emit("poll", pollId, new Date().getTime());
+
+        // Find all transaction hashes we are waiting on
+        this._events.forEach((event) => {
+            switch (event.type) {
+                case "tx": {
+                    const hash = event.hash;
+                    let runner = this.getTransactionReceipt(hash).then((receipt) => {
+                        if (!receipt) { return null; }
+                        this.emit(hash, receipt);
+                        return null;
+                    }).catch((error: Error) => { this.emit("error", error); });
+
+                    runners.push(runner);
+
+                    break;
+                }
+
+                case "filter": {
+                    const filter = event.filter;
+                    // ensure we don't get from == to
+                    const from = this._previousToTimestamp || new Date().getTime() - this._pollingInterval;
+                    const to = new Date().getTime();
+                    this._previousToTimestamp = to;
```

Comment:
Is the query for logs `inclusive` or `exclusive` for `from` & `to`? If they are both `inclusive`, we should make sure that `to` is not equal to `from`